Missing addon: GW2 UI and others

**Describe the bug**
A list of missing addons from my setup

**To Reproduce**
Steps to reproduce the behavior:
1. Install the missing addons marked in the screenshot

**Expected behavior**
The addons should appear in ajour

**Screenshots**

![image](https://user-images.githubusercontent.com/7250921/94586211-406f3a00-0281-11eb-9f59-3a90ed0552bd.png)

**Software involved**
Please complete the following information:
 - OS: Pop_OS! 20.04 (Ubuntu)
 - Ajour version: 0.3.4 Installed manually as described at https://github.com/casperstorm/ajour
 - Missing addons: Guild Wars 2 UI, masque_shadow, ouf_hank, ouf_ruri

**Additional context**
I am not sure if i got the ouf addons from curse but i think so. concerning guild wars 2 ui: this is 100% from the curse website

I tried this and the follow is a problem:
- Masque_Shadow
- GW2_UI
- oUF_Hank


This is the log error:
```
Finished dev [unoptimized Finished dev [unoptimized + debuginfo] target(s) in 0.29s
08:25:07.542 [ajour_core::parse][ERROR] fingerprinting failed for "/Applications/World of Warcraft/_retail_/Interface/AddOns/oUF_Hank": Unable to find file: "/Applications/World of Warcraft/_retail_/Interface/AddOns/oUF_Hank/config.lua"
08:25:07.543 [ajour_core::parse][ERROR] fingerprinting failed for "/Applications/World of Warcraft/_retail_/Interface/AddOns/Masque_Shadow": stream did not contain valid UTF-8
08:25:07.678 [ajour_core::parse][ERROR] fingerprinting failed for "/Applications/World of Warcraft/_retail_/Interface/AddOns/GW2_UI": Unable to find file: "/Applications/World of Warcraft/_retail_/Interface/AddOns/GW2_UI/libs/libs/LibStub/LibStub.lua"
```
Issue with `Masque_Shadow` is the authors name is not UTF-8 valid: `## Author: Fedorenko R.D. √Ä√¥√•√¨-√ó√•√∞√≠√ª√©√ò√∞√†√¨`.

Issue with `oUF_Hank` is that we try to find `/oUF_Hank/config.lua` but its called `/oUF_Hank/config.default.lua`
Issue with `GW2_UI` is that we try to find `/GW2_UI/libs/LibStub/LibStub.lua` but its located here `/GW2_UI/libs/libs/LibStub/LibStub.lua`

I think with this information we have enough to fix these bugs. Thanks for reporting @moggybaer.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 The sections should follow the order `Packaging`, `Added`, `Changed`, `Fixed` and `Removed`.
 
 ## [Unreleased]
+### Fixed
+- Fixed fingerprinting where some addons would fail during fingerprinting due to invalid UTF-8 characters and missing files. These addons now successfully fingerprint.
 
 ## [0.4.0] - 2020-10-6
 ### Added

--- a/crates/core/src/parse.rs
+++ b/crates/core/src/parse.rs
@@ -642,17 +642,13 @@ pub fn fingerprint_addon_dir(
                     .as_str();
                 // Path might be case insensitive and have windows separators. Find it
                 let path_match = path_match.replace("\\", "/");
-                let parent = path.parent().ok_or_else(|| {
-                    ClientError::FingerprintError(format!("No parent directory for {:?}", path))
-                })?;
-                let file_to_find = parent.join(Path::new(&path_match));
-                let real_path = find_file(&file_to_find).ok_or_else(|| {
-                    ClientError::FingerprintError(format!(
-                        "Unable to find file: {:?}",
-                        file_to_find
-                    ))
-                })?;
-                to_parse.push_back(real_path);
+                if let Some(parent) = path.parent() {
+                    let file_to_find = parent.join(Path::new(&path_match));
+
+                    if let Some(real_path) = find_file(&file_to_find) {
+                        to_parse.push_back(real_path);
+                    }
+                }
             }
         }
     }

--- a/crates/core/src/parse.rs
+++ b/crates/core/src/parse.rs
@@ -17,7 +17,7 @@ use rayon::prelude::*;
 use serde::{Deserialize, Serialize};
 use std::collections::{HashMap, HashSet, VecDeque};
 use std::fs::File;
-use std::io::{BufRead, BufReader};
+use std::io::{BufRead, BufReader, Read};
 use std::path::{Path, PathBuf};
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::time::SystemTime;
@@ -613,7 +613,13 @@ pub fn fingerprint_addon_dir(
             file_parsing_regex.get(&ext).ok_or_else(|| {
                 ClientError::FingerprintError(format!("ext not in file parsing regex: {:?}", ext))
             })?;
-        let text = std::fs::read_to_string(&path).map_err(ClientError::fingerprint)?;
+        let mut file = File::open(&path).map_err(ClientError::fingerprint)?;
+
+        let mut buf = vec![];
+        file.read_to_end(&mut buf)
+            .map_err(ClientError::fingerprint)?;
+
+        let text = String::from_utf8_lossy(&buf);
         let text = comment_strip_regex.replace_all(&text, "");
         for line in text.split(&['\n', '\r'][..]) {
             let mut last_offset = 0;

